terraso_backend/apps/collaboration/models/memberships.py

```diff
@@ -0,0 +1,146 @@
+        (ENROLL_METHOD_INVITE, _("Invite")),
+        (ENROLL_METHOD_BOTH, _("Both")),
+    )
+
+    enroll_method = models.CharField(
+        max_length=10, choices=ENROLL_METHODS, default=DEFAULT_ENROLL_METHOD_TYPE
+    )
+
+    members = models.ManyToManyField(User, through="Membership")
+
+    membership_type = models.CharField(
+        max_length=32,
+        choices=MEMBERSHIP_TYPES,
+        default=DEFAULT_MEMERBSHIP_TYPE,
+    )
+
+    def save_member(self, kwargs):
+        user = kwargs["user"]
+        user_role = kwargs["user_role"]
+        membership_status = kwargs["membership_status"]
+
+        membership = self.get_membership(user)
+        is_closed = self.membership_type == MembershipList.MEMBERSHIP_TYPE_CLOSED
+        is_new = not membership
+        if is_new:
+            membership = Membership(
+                membership_list=self,
+                user=user,
+                user_role=user_role,
+                membership_status=Membership.PENDING if is_closed else Membership.APPROVED,
+            )
+        else:
+            previous_membership_status = membership.membership_status
+            membership.user_role = user_role
+            if membership_status:
+                membership.membership_status = Membership.get_membership_status_from_text(
+                    membership_status
+                )
+
+        membership_status = membership.membership_status
+
+        is_membership_approved = (
+            previous_membership_status != Membership.APPROVED
+            and membership_status == Membership.APPROVED
+        )
+
+        if is_membership_approved:
+            EmailNotification.send_membership_approval(membership.user, membership.membership_list)
+
+        membership.save()
+        return membership
+
+    @property
+    def members(self):
+        member_memberships = models.Subquery(self.memberships.approved_only().values("user_id"))
+        return User.objects.filter(id__in=member_memberships)
+
+    def has_role(self, user: User, role: str) -> bool:
+        return self.memberships.by_role(role).filter(id=user.id).exists()
+
+    def is_member(self, user: User) -> bool:
+        return self.members.filter(id=user.id).exists()
+
+    def get_membership(self, user: User):
+        return self.memberships.filter(user=user).first()
+
+    @property
+    def can_join(self):
+        return self.enroll_method in (self.ENROLL_METHOD_JOIN, self.ENROLL_METHOD_BOTH)
+
+    @classmethod
+    def get_membership_type_from_text(cls, membership_type):
+        if membership_type and membership_type.lower() == cls.MEMBERSHIP_TYPE_CLOSED:
+            return cls.MEMBERSHIP_TYPE_CLOSED
+        return cls.MEMBERSHIP_TYPE_OPEN
+
+
+class MembershipObjectsManager(SafeDeleteManager):
+    def by_role(self, role):
+        return self.filter(user_role=role, membership_status=Membership.APPROVED)
+
+    def approved_only(self):
+        return self.filter(membership_status=Membership.APPROVED)
```

Comment:
question: can we reuse `approved_only` from groups.py?

Reply:
@paulschreiber My plan is to handle this on separate stages, the plan is to eventually use the new membershipList model to get the functionality of members to Groups, Landscapes and Projects too. that is why right now there is a lot of duplication between Group and MemberhipList, but the second stage after the membership list for story maps is approved will be to connect the other 3 entities to member list and remove the membership logic coupled with group right now